### ISSUE TITLE
TypeError: Incorrect padding

### ISSUE THREAD

**Issue by bitstreamout** [unode/firefox_decrypt#40]
On openSUSE Leap 15.0 (kernel 4.12.14, python 2.7.14) I see during command

    python bin/firefox_decrypt -c 1  --export-pass

a lot of Website, Username, and Password entries but then it fails suddenly with

```
Traceback (most recent call last):
  File "/suse/werner/bin/firefox_decrypt", line 965, in <module>
    main()
  File "/suse/werner/bin/firefox_decrypt", line 946, in main
    csv_quotechar=args.quotechar,
  File "/suse/werner/bin/firefox_decrypt", line 536, in decrypt_passwords
    user, passw = self.decode_entry(user, passw)
  File "/suse/werner/bin/firefox_decrypt", line 503, in decode_entry
    passw = self.NSS.decode(passw64)
  File "/suse/werner/bin/firefox_decrypt", line 400, in decode
    data = b64decode(data64)
  File "/usr/lib64/python2.7/base64.py", line 78, in b64decode
    raise TypeError(msg)
TypeError: Incorrect padding
```

I guess that as my password are autogenerated not only with characters and numbers but also with special signs, that the script is running onto CVS quote/delimeter character

**Comment by unode** [comment]
What version of firefox_decrypt did you use?
If latest from git, there are some known issues with it that I haven't been able to fix yet (time constraints).

Can you test with the latest release (0.7.0) and see if the same happens?

**Comment by bitstreamout** [comment]
AFAICS this is 0.7.0 ...
`
bin/firefox_decrypt --version
`
`
0.7.0
`

**Comment by bitstreamout** [comment]
I had downloaded
`https://raw.githubusercontent.com/unode/firefox_decrypt/master/firefox_decrypt.py`
and just now retried this and executed
`diff -up firefox_decrypt.py bin/firefox_decrypt`
no difference

**Comment by unode** [comment]
Let me rephrase. Can you try https://github.com/unode/firefox_decrypt/releases/tag/0.7.0 instead?

`master` is the development version and is currently not stable.

**Comment by bitstreamout** [comment]
Similar result only the line numbers do change:

`Traceback (most recent call last):`
`  File "/suse/werner/bin/firefox_decrypt", line 861, in <module>`
`    main()`
`  File "/suse/werner/bin/firefox_decrypt", line 849, in main`
`    csv_quotechar=args.quotechar,`
`  File "/suse/werner/bin/firefox_decrypt", line 449, in decrypt_passwords`
`    user, passw = self.decode_entry(user, passw)`
`  File "/suse/werner/bin/firefox_decrypt", line 410, in decode_entry`
`    passw = self.NSS.decode(passw64)`
`  File "/suse/werner/bin/firefox_decrypt", line 309, in decode`
`    data = b64decode(data64)`
`  File "/usr/lib64/python2.7/base64.py", line 78, in b64decode`
`    raise TypeError(msg)`
`TypeError: Incorrect padding`

**Comment by unode** [comment]
Interesting! Don't think I've ever encountered this particular error.
I would love to have this as a test-case. 

Can you reproduce this if you create a new Firefox profile and add the password that is currently failing? If yes, could you attach that Firefox profile here?

Thanks!

**Comment by bitstreamout** [comment]
I found the culprit ... there is an old entry which does not show a base64 encrypted passphrase but a simply string (could be an old entry done by a very old version of 'Secure Password Editor' I had often used to remember  ... besides password ... other data for some sides) ... I've changed the python script in `decode_entry()` to catch the exception, my code now looks like

```python
    def decode_entry(self, user64, passw64):
        """Decrypt one entry in the database
        """
        LOG.debug("Decrypting username data '%s'", user64)
        try:
            user = self.NSS.decode(user64)
        except Exception:
            user = user64

        LOG.debug("Decrypting password data '%s'", passw64)
        try:
            passw = self.NSS.decode(passw64)
        except Exception:
            passw = passw64

        return user, passw
```

**Comment by unode** [comment]
I'm not confident in using those exception handlers. They assume that any decoding error is fixable by simply using the raw record, which is true in your case but false in any other scenario.

On the other hand, supporting records created by third party tools that do not follow the specification is beyond scope. I'll consider the scenario for a test-case but will implement a different solution that simply skips bad records.

**Comment by kastixx** [comment]
I've stumbled upon the similar problem. It seems that it is somewhat connected to using null-terminated string type for binary data in `class SECItem`. Replacing it with `POINTER(c_char)` apparently fixes this.
